compiler/modules/front/resolveTypes.bal

```diff
@@ -28,6 +28,42 @@ function resolveTypes(ModuleSymbols mod) returns ResolveTypeError? {
     foreach var defn in mod.defns {
         check resolveDefn(mod, defn);
     }
+    check finishDeferredEmptinessChecks(mod);
+}
+
+function finishDeferredEmptinessChecks(ModuleSymbols mod) returns ResolveTypeError? {
+    if !mod.tc.env.isReady() {
+        panic err:impossible("type environment is not ready");
+    }
+    foreach var { semType, modDefn, td } in mod.deferredEmptinessChecks {
+        // XXX When we can give multiple errors we should check all the deferred emptiness checks
+        check finishEmptinessCheck(mod, semType, modDefn, td);
+    }
+    foreach var { startPos, endPos, modDefn, recursive } in mod.atomicEmptyTypeRecursionCheck {
+        if !recursive {
+            d:Location loc = s:locationInDefn(modDefn, { startPos, endPos });
+            return err:semantic("non recursive empty type", loc);
+        }
+    }
+}
+
+function finishEmptinessCheck(ModuleSymbols mod, t:SemType semType, s:ModuleLevelDefn modDefn, s:TypeDesc td) returns ResolveTypeError? {
+    if !t:isEmpty(mod.tc, semType) {
+        return;
+    }
+    d:Position startPos = td.startPos;
+    d:Position endPos = td.endPos;
+    d:Location loc = s:locationInDefn(modDefn, { startPos, endPos });
+    if td is s:BinaryTypeDesc && td.op is "&" {
+        return err:semantic("intersection must not be empty", loc);
+    }
+    if t:isSemTypeRecursive(semType) {
```

Comment:
I don't understand why you need the `atomicEmptyTypeRecursionCheck` table.

What I suggested was not to give an error right away for a non-recursive atomic empty type, because there will should be an empty intersection or recursive atomic type that is the real cause of the problem.  So we just need to save one of the non-recursive empty types that we can use to give a panic in case (unexpectedly) we don't find a recursive error.  This also requires catching a member type for which isNever is true.

Reply:
fixed

Reply:
fixed